### PR TITLE
Added gate cancellation optimisation with relevant test

### PR DESCRIPTION
**Pull request by WashimNeupane**
Simple Hadamard cancellation (H followed by another H)
Simple PauliX cancellation (X followed by X). 
Pauli cancellation is in template format so can easily add Y and ZOp when needed.

### CODE REVIEW

File: quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp [hunk 1]
```diff
@@ -32,53 +33,57 @@ struct QuantumOptimisePass
...
-//     return success();
-//   }
-// };
+struct CancelHadamardPairs : public OpRewritePattern<HOp> {
+    using OpRewritePattern<HOp>::OpRewritePattern;
+
+    LogicalResult
+    matchAndRewrite(HOp op, PatternRewriter &rewriter) const override
+    {
+        Operation* nextOp = op->getNextNode();
```
Comment (lschuetze) [inline quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp:42]: I think this is the wrong API. You are currently requiring that gates have to follow op by op. However, there might be gates in between that do not work on the same values, i.e., the constraint is that the produced value of the `HOp` we currently visit is consumed by another `HOp`.

Comment (lschuetze) [inline quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp:42]: You should check that you are working on single-qubit gates. The multi-qubit to single-qubit transformation pass must run before.

File: quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp [hunk 2]
```diff
@@ -32,53 +33,57 @@ struct QuantumOptimisePass
...
+        if (nextOp && isa<HOp>(nextOp)) {
+            HOp nextHadamard = cast<HOp>(nextOp);
```
Comment (lschuetze) [inline quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp:44]: Please prefer `llvm::dyn_cast`

File: quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp [hunk 3]
```diff
@@ -88,9 +93,8 @@ void QuantumOptimisePass::runOnOperation()
 
 void mlir::quantum::populateQuantumOptimisePatterns(RewritePatternSet &patterns)
 {
-    // patterns.add<
-    //   HermitianCancel,
-    //   AdjointCancel>(patterns.getContext());
+    patterns.add<CancelHadamardPairs>(patterns.getContext());
+    patterns.add<CancelPauliPairs<XOp>>(patterns.getContext());
```
Comment (lschuetze) [inline quantum/lib/Dialect/Quantum/Transforms/Optimise.cpp:97]: Please use the `patterns.add<ALL_PATTERNS_HERE>(...)`

File: quantum/test/Dialect/Quantum/Transforms/test_quantum_optimise.mlir [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+module {
+  // CHECK-LABEL: func.func @main(
+  func.func @main() -> !quantum.qubit<1> {
+    // CHECK-DAG: %[[Q:.+]] = "quantum.alloc"() : () -> !quantum.qubit<1>
+    %q = "quantum.alloc"() : () -> (!quantum.qubit<1>)
+    %q1 = "quantum.H" (%q) : (!quantum.qubit<1>) -> (!quantum.qubit<1>)
+    %q2 = "quantum.H" (%q1) : (!quantum.qubit<1>) -> (!quantum.qubit<1>)
+    %q3 = "quantum.X" (%q2) : (!quantum.qubit<1>) -> (!quantum.qubit<1>)
+    %q4 = "quantum.X" (%q3) : (!quantum.qubit<1>) -> (!quantum.qubit<1>)
+    // CHECK-DAG: return %[[Q]] : !quantum.qubit<1>
```
Comment (lschuetze) [inline quantum/test/Dialect/Quantum/Transforms/test_quantum_optimise.mlir:12]: Fix the usage of LIT checks.

File: quantum/test/Dialect/Quantum/Transforms/test_quantum_optimise.mlir [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// RUN: quantum-opt --quantum-optimise %s | FileCheck %s
```
Comment (lschuetze) [inline quantum/test/Dialect/Quantum/Transforms/test_quantum_optimise.mlir:1]: Add more tests. Think about edge cases. Make your tests as small as possible and provide a good name, e.g., `func.func @cancel_consecutive_hops() -> ()`
